tools/topsites.py

```diff
@@ -0,0 +1,182 @@
+Session = sessionmaker(bind=engine)
+session = Session()
+
+
+class Site(Base):
+    """SQLAchemy base object for an Alexa top site."""
+    __tablename__ = "topsites"
+
+    url = Column(String, primary_key=True)
+    priority = Column(Integer)
+    country_code = Column(String)
+    ranking = Column(Integer)
+
+    def __init__(self, url, priority, country_code, ranking):
+        self.url = url
+        self.priority = priority
+        self.country_code = country_code
+        self.ranking = ranking
+
+Base.metadata.create_all(engine)
+
+
+def query_topsites(country_code, count=1000):
+    """Query top sites with given country code and count."""
+    for index in range(1, count, 100):
+        uri = build_uri(country_code, index)
+        response = requests.get(uri)
+
+        if response.status_code == 200:
+            # See http://docs.aws.amazon.com/AlexaTopSites/latest/index.html?QUERY_QueryRequests.html  # nopep8
+            dom = parseString(response.content)
+            sites = dom.getElementsByTagName("aws:Site")
+            for site in sites:
+                parse_site(site)
+            session.commit()
+
+
+def parse_site(site):
+    """Parse given dom object."""
+    url = node_text(site, 'aws:DataUrl')
+    rank = int(node_text(site, 'aws:Rank'))
+
+    # Global website, initial priority is 1
+    # Tier 1 regions, initial priority is 2
+    priority = 1 if country_code == "" else 2
+
+    # Ranking between 100 and 1000, add priority 1
+    # Ranking between 1000 and 10000, add priority 2
+    if rank > 100 and rank <= 1000:
+        priority = priority + 1
+    elif rank > 1000 and rank <= 10000:
+        priority = priority + 2
+
+    if url not in topsites:
+        # URL not cached, create Site object and put in topsites
+        site_row = Site(url, priority, country_code, rank)
+        topsites[url] = site_row
+        session.add(site_row)
+    else:
+        site_row = topsites[url]
+        # If priority of the URL is higher than cached one,
+        # update new priority, country_code and ranking in cache
+        if site_row.priority > priority:
+            site_row.priority = priority
+            site_row.country_code = country_code
+            site_row.ranking = rank
+
+
+def build_uri(country_code, start_ranking):
+    """Build Alexa top site URI with given country code and start ranking."""
+    # Build query string
+    query_string = build_query_string(country_code, start_ranking)
+
+    # String to sign
+    to_sign = "GET\n" + ATS_SERVICE_HOST + "\n/\n" + query_string
+    signature = gen_sign(to_sign)
+
+    # URI with signature
+    uri = "{base}{query}&{signature}".format(
+        base=ATS_AWS_BASE_URL,
+        query=query_string,
+        signature=urlencode({"Signature": signature}, "UTF-8"))
+
+    return uri
+
+
+def build_query_string(country_code, start_ranking, timestamp=None):
+    """Build query string for request with start ranking and count."""
+    if None == timestamp:
```

Comment:
This is the opposite.

```python
if timestamp == None:
```

but more to the point. Are there anytime where you do not need timestamp to build the query string. 
If it has always to be there. drop it from the parameter and just call the timestamp.

```python
def build_query_string(country_code, start_ranking):
    """Build query string for request with start ranking and count."""
    nowtime = datetime.datetime.utcnow()
    timestamp = '{}Z'.format(nowtime.strftime('ATS_DATEFORMAT')[:-3])
    # Alexa top site only accept request with ordered query parameters
    # … etc.
```